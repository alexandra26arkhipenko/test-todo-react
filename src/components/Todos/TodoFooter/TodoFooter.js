import { useSelector } from "react-redux";

import {
    TODO_FOOTER_ITEMS_LEFT,
    TODO_FOOTER_WELL_DONE,
  } from "../../../constants/constants";

import classes from "./TodoFooter.module.css";
import TodoFilter from "./TodoFilter/TodoFilter";

const TodoFooter = () => {
  const todos = useSelector((state) => state.todos);

  const spanValue =
    todos.filter((x) => !x.isComplited).length === 0
      ? TODO_FOOTER_WELL_DONE
      : `${
          todos.filter((x) => !x.isComplited).length
        } ${TODO_FOOTER_ITEMS_LEFT}`;

  return (
    <div className={classes.footer}>
      <span>{spanValue}</span>
      <TodoFilter />
    </div>
  );
};

export default TodoFooter;
