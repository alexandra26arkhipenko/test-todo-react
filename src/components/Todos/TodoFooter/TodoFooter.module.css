.footer {
    outline: none;
    color: #d7cac1db; 
    margin: 20px;
    margin-bottom: 20px;
    margin-left: 35px;
    align-items: center;
    display: flex;
    justify-content: space-between;
    padding-right: 13px;
    font-size: 15px;
}

.footer button:hover {
    color: rgb(114 12 12);
}