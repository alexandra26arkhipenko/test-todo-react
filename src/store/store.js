import { configureStore } from "@reduxjs/toolkit";
import { localStorageMiddleware } from "./middleware/localStorageMiddleware";

import todoReducer from "./todo";

const store = configureStore({
  reducer: todoReducer,
  middleware: [localStorageMiddleware],
});

export default store;
