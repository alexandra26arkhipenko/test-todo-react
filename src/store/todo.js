import { createSlice } from "@reduxjs/toolkit";
import { v4 as uuid } from "uuid";
import { getItemFromLocalStorage } from "../repository/localStorageRepository";
import { TODO_FOOTER_FILTER_ALL } from "../constants/constants";

const savedTodos = getItemFromLocalStorage("todos");
let todos = [];

if (savedTodos) {
  todos = JSON.parse(savedTodos);
}

const initialState = {
  todos: todos,
  filter: "",
};

const todoSlice = createSlice({
  name: "counter",
  initialState,
  reducers: {
    addTodo(state, action) {
      const newTodo = {
        id: uuid(),
        text: action.payload,
        isCompleted: false,
      };

      state.todos.unshift(newTodo);
    },
    compliteTodo(state, action) {
      const index = state.todos.findIndex((todo) => todo.id === action.payload);

      if (index !== -1) {
        state.todos[index].isCompleted = !state.todos[index].isCompleted;
      }
    },
    deleteTodo(state, action) {
      const index = state.todos.findIndex((todo) => todo.id === action.payload);

      if (index !== -1) {
        state.todos.splice(index, 1);
      }
    },
    deleteAllCompletedTodos(state) {
      state.todos = state.todos.filter((todo) => !todo.isCompleted);
    },
    filterTodos(state, action) {
      state.filter = action.payload;
    },
    reorderTodos(state, action) {
      let { dragIndex, dropIndex, filteredTodos } = action.payload;

      if (state.filterValue !== TODO_FOOTER_FILTER_ALL) {
        dragIndex = state.todos.findIndex(
          (property) => property.id === filteredTodos[dragIndex].id
        );
        dropIndex = state.todos.findIndex(
          (property) => property.id === filteredTodos[dropIndex].id
        );
      }

      const updatedTodos = [...state.todos];
      const [reorderedItem] = updatedTodos.splice(dragIndex, 1);
      updatedTodos.splice(dropIndex, 0, reorderedItem);
      state.todos = updatedTodos;
    },
  },
});

export const todoActions = todoSlice.actions;
export default todoSlice.reducer;
