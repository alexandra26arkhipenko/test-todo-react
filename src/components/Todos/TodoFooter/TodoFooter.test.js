import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import TodoFooter from "./TodoFooter";
import { todoActions } from "../../../store/todo";
import configureMockStore from "redux-mock-store";

const props = {
  onFilter: jest.fn(),
  activeTodosCount: 1,
};
const mockStore = configureMockStore();
const store = mockStore({});

describe("TodoFooter", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <TodoFooter {...props} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should delete all completed when delete all btn is clicked", () => {
    render(
      <Provider store={store}>
        <TodoFooter {...props} />
      </Provider>
    );

    fireEvent.click(screen.getByText("Clear Completed"));

    expect(store.getActions()).toContainEqual(
      todoActions.deleteAllCompletedTodos()
    );
  });
});
