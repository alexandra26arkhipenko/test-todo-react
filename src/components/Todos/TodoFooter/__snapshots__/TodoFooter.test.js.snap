// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFooter should render correctly 1`] = `
<div>
  <div
    class="footer"
  >
    <span>
      1 items left
    </span>
    <div>
      <button
        class="button false"
        type="button"
        value="All"
      >
        All
      </button>
      <button
        class="button false"
        type="button"
        value="Active"
      >
        Active
      </button>
      <button
        class="button false"
        type="button"
        value="Completed"
      >
        Completed
      </button>
    </div>
    <button
      class="button undefined"
      type="button"
      value="Clear Completed"
    >
      Clear Completed
    </button>
  </div>
</div>
`;
